### ISSUE TITLE
A few array related type issues.

### ISSUE THREAD

**Issue by tanzanite777** [DEVSENSE/phptools-docs#479]
Type: <b>Bug</b>

Ran into weird type issues that took some head-scratching / investigating - culminating with the following reduced (well, went a bit broader during investigation) demo-class.

```php
class Test {
    /** @var Test[][] Fishy */
    protected array $foo = []; // [issue] Hover $foo: nothing shows up - no type nor description "Fishy"


    function bar() : void {
        $foo = $this->foo; // [issue] _IF_ 'snafu()' function exists then left side type hint will be 
                           // 'array<array<Test|array>>' instead of the 'Test[][]' on the right side.
                           // [issue] Cause seems to be type inference from 'snafu()' code - which is very useful
                           // in principle, but should not affect property type if it has already
                           // been explicitly declared ("/** @var Test[][] fishy */").
        ksort($foo); // [issue] type hint for $foo transforms into 'array<array<Test|array>, mixed>' ... mmm, what? :)
    }

    /** @return $this [feature] Sneaky unrelated plea for consideration: '$this' should show up as return type in description (currently it gets removed). https://wiki.php.net/rfc/this_return_type */
    function snafu(Test|array|null $obj) : static {
        if($obj === null) return $this;
        // Type hint has correctly lost 'null' from hereon
        if(is_array($obj)) {
            foreach($obj as $one) if($one instanceof Test) $one->doNothing(); // all good
        } else {
            $id = spl_object_id($obj); // [issue] Type hint for $obj is 'Test|array' - but it cannot be array
            $this->foo[0][$id] = $obj; // Poisoning: (incorrect) type inference overrides correct type elsewhere - see: 'bar()'
        }
        return $this;
    }


    /**
     * not fishy
     * @return Test[][]
     */
    function foo() : array { // [issue] 'foo()' has type hint of 'array' instead of 'Test[][]'
        return $this->foo; // [feature?] I wonder if it would be worth the trouble of giving both 'foo' options their own description section (instead of just showing the description for the last one)?
    }

    function goodButWeird() : void {
        // Weird: 'foo()' has imprecise type hint of 'array', but $set / $one get the correct 'Test[]' / 'Test'
        // Seems that type propagates correctly - just that hover hint for 'foo()' ends up hiding it
        foreach($this->foo() as $set) foreach($set as $one) $one->doNothing();
    }

    function doNothing() : void {}
}
```

Extension version: 1.43.14858
VS Code version: Code 1.85.2 (8b3775030ed1a69b13e4f4c628c612102e30a681, 2024-01-18T06:40:10.514Z)
OS version: Windows_NT x64 10.0.19045
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz (8 x 3600)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|undefined|
|Memory (System)|15.88GB (5.31GB free)|
|Process Argv|--crash-reporter-id adf79977-8c83-4303-918e-a7dfc5f31bf1|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vspor879:30202332
vspor708:30202333
vspor363:30204092
vscod805cf:30301675
binariesv615:30325510
vsaa593:30376534
py29gd2263:30899288
c4g48928:30535728
azure-dev_surveyone:30548225
a9j8j154:30646983
0bi6i642:30951795
pythongtdpath:30769146
welcomedialog:30910333
pythonidxpt:30866567
pythonnoceb:30805159
asynctok:30898717
pythontestfixt:30902429
pythonregdiag2:30936856
pyreplss1:30897532
pythonmypyd1:30879173
pythoncet0:30885854
pythontbext0:30879054
dsvsc016:30899300
dsvsc017:30899301
dsvsc018:30899302
dsvsc019bcf:30953938
3ef8e399:30949928

```

</details>

**Comment by tanzanite777** [comment]
Sorry about piling issues into one single report. Things went a bit out of hand when i tried to track down my original issue (type issues where i suspected "poisoning from type inference" but had trouble confirming / locating the thing) and turned out it was caused by multiple separate issues working together.

Reading my report, with a clearer head the next day, some clarifications are in order for one of the issues:
```
class Test2 {
    /**
     * @param array<mixed, array<mixed, Test2>> $a Dummy description (just sanity check for key/value type order).
     * @param Test2[][] $b
     * @param Test2 $c Dummy description.
     * @return Test2[][]
     */
    function foo(array $a, array $b, $c) : array {
        // Hovering $a, $b, $c i see that all types are fully qualified.
        // However, hovering 'foo' shows a hint that has lost nearly all of the type info 
        // ($c being the sole exception - it's basic PHP type). This is a problem ... see 'bar()'
        return $a + $b;
    }

    function bar(array $x) : array {
        return $this->foo($x, $x, $this);
        // Writing the previous line i get only "foo(array $a, array $b, Test2 $c): array<array<Test2>>" for type hints.
        // Seeing what type is expected / allowed is invaluable - currently the only way to see it is to go read the original PHPdoc in source form.
        // Surprisingly i do get to see the return type (in its, in this case unnecessarily, detailed/verbose form) - however, all other types are gone.
        // Note also that the, surprisingly present, return type will be gone when just trying to read the code and hovering 'foo(...)' for details.
    }
}
```

It would be nice if this could be solved somehow, but it probably is not an easy task - might have some gotchas or just GUI burdening issues or whatnot.

I, personally, would prefer function / property signatures (*) to always show the fully qualified types (not limited to existing PHP type system which is based on enforceability and not the real world needs). I think that would be a good solution for everyone as the underlying enforceable type is implicit anyway (except when enforceable type was not specified to begin with - but that caveat is currently not adhered to anyway, see $c in 'foo()' where true signature is uenforceable/'mixed' but 'Test' is shown instead). Well, at least i cannot think of any drawbacks atm.

Alternative, which i would not particularly like, would be to show full type in @param/@return.

*) Hopefully it is obvious what i mean by 'signature' - not sure what one calls thous in PHP land.

----------
You probably will notice this anyway, but probably wont hurt to draw attention as it was not specifically mentioned in your reply (sorry again for piling issues into one report x_x). In OP:
* 'is_array($obj)' based branch did not remove array type from $obj in the else branch.
* ksort somehow managed to get key type 'array<Test|array>' - perhaps worth adding failsafe/reporting/asserts to block impossible array keys showing up regardless of source (ksort of anything else).

**Comment by jakubmisek** [comment]
I'll fix the tooltips for function signatures, and we'll see what we can improve next.

**Comment by jakubmisek** [comment]
Pre-release `1.44.14939` improves tooltips, `ksort`, `is_array`, and array type hints. Hover tooltips should reveal more info now.

**Comment by jakubmisek** [comment]
Released `1.44.14950` - improves `is_array` check, and `array_rand`

**Comment by jakubmisek** [comment]
Closing as it is solved, please comment if there is more. Thanks!
